### PR TITLE
Add an alternative to torch.nn.parallel.DistributedDataParallel

### PR DESCRIPTION
**Pull request by HiroakiMikami**
Add a module for data parallelism to handle `torch.util.checkpoint` and dynamic computational graph correctly.

### CODE REVIEW

**Review by emcastillo — APPROVED** [review]
This looks awesome

File: tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    def forward(self, x):
+        y = self.l0(x)
+        y = checkpoint(self.l1, y)
+        y = checkpoint(self.l2, y)
+        return y
+
+
+def _run(init_file, input, module, rank, args, step, device_type):
+    dist.init_process_group(backend="gloo", init_method=f"file://{init_file}",
+                            world_size=2, rank=rank)
```
Comment (emcastillo) [inline tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: We need to ensure that there are 2 gpus in the tests, sorry I overlooked this.

Comment (HiroakiMikami) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: I think these tests require only 1 gpu.
It's true that I used 2 processes for these tests. But both processes use same GPUs: https://github.com/pfnet/pytorch-pfn-extras/pull/80/files/7386f3a7bdeab2d5b5676172d1dda0e54cd8db39#diff-336610f8599a77e118a5a09f91cbefc98c62f66e9928cd98927abebd8f4b8694R90

Comment (emcastillo) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: ah yes, we are not setting a local device per process :), but I think its better if we do so

Comment (HiroakiMikami) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: I set the local device for each process to `cuda:0` in line 90 of `test_distributed.py`:

```python
    elif device_type == "cuda":
        device = torch.device(f"{device_type}:0")
```

Comment (emcastillo) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: Can we use two gpus instead? we have modified the CI settings to allocate 2 gpus, so it should test this under similar to real-world conditions :)

Comment (HiroakiMikami) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: ok, I'm changing the test script now.

Comment (HiroakiMikami) [reply tests/pytorch_pfn_extras_tests/nn_tests/parallel_tests/test_distributed.py:86]: I modified the test cases for using 2 GPUs.

**Pull request merged**